lib/fir/statements.h

```diff
@@ -0,0 +1,231 @@
+// Copyright (c) 2018, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef FORTRAN_IR_STATEMENTS_H_
+#define FORTRAN_IR_STATEMENTS_H_
+
+#include "mixin.h"
+#include "../common/idioms.h"
+#include "../common/indirection.h"
+#include "../evaluate/expression.h"
+#include "../parser/parse-tree.h"
+#include "../semantics/symbol.h"
+#include <initializer_list>
+#include <iostream>
+
+namespace Fortran::IR {
+
+class Statement;
+class BasicBlock;
+CLASS_TRAIT(StatementTrait)
+CLASS_TRAIT(TerminatorTrait)
+CLASS_TRAIT(ActionTrait)
+
+using Expression = const common::OwningPointer<evaluate::GenericExprWrapper>;
+using Variable = const semantics::Symbol;
+using PathVariable = const parser::Variable;
+
+struct Evaluation
+  : public SumTypeCopyMixin<
+        std::variant<Expression *, Variable *, PathVariable *>> {
+  SUM_TYPE_COPY_MIXIN(Evaluation)
+  Evaluation(PathVariable *pv) : SumTypeCopyMixin{pv};
+  template<typename A> Evaluation(A *a) : SumTypeCopyMixin{a} {}
+  std::string dump() const;
+};
+
+struct Stmt_impl {
+  using StatementTrait = std::true_type;
+};
+
+struct TerminatorStmt_impl : public Stmt_impl {
+  virtual std::list<BasicBlock *> succ_blocks() const;
+  using TerminatorTrait = std::true_type;
+};
+
+struct ReturnStmt : public TerminatorStmt_impl {
+  ~ReturnStmt() = default;
+  static ReturnStmt *Create();
+  static ReturnStmt *Create(Expression *expression);
+
+private:
+  Expression *returnValue;
+  explicit ReturnStmt(Expression *);
+};
+
+struct BranchStmt : public TerminatorStmt_impl {
+  friend Statement;
+  ~BranchStmt() = default;
+  static BranchStmt *Create(
+      Expression *condition, BasicBlock *trueBlock, BasicBlock *falseBlock);
+  static BranchStmt *Create(BasicBlock *succ);
+  bool hasCondition() const;
+  std::list<BasicBlock *> succ_blocks() const override;
+
+private:
+  explicit BranchStmt(
+      Expression *condition, BasicBlock *trueBlock, BasicBlock *falseBlock);
+  static constexpr unsigned TrueIndex{0u};
+  static constexpr unsigned FalseIndex{1u};
+  Expression *condition;
+  BasicBlock *succs[2];
+};
+
+struct SwitchStmt : public TerminatorStmt_impl {
+  using ValueSuccPairType = std::pair<Expression *, BasicBlock *>;
+  using ValueSuccPairListType = std::vector<ValueSuccPairType>;
+  ~SwitchStmt() = default;
```

Comment:
Why the explicitly defaulted destructors?  Aren't the destructors defaulted anyway?

Reply:
I honestly don't recall now, but I might have picked them up from the SIL code? True, they don't appear to be useful here.